src/rdf4cpp/Node.cpp

```diff
@@ -120,24 +120,82 @@ bool Node::is_inlined() const noexcept {
-        return std::strong_ordering::equivalent;
+TriBool Node::eq_impl(Node const &other) const noexcept {
+    if (null() || other.null()) {
+        // "Apart from BOUND, COALESCE, NOT EXISTS and EXISTS, all functions and operators operate on RDF Terms and will produce a type error if any arguments are unbound."
+        // - https://www.w3.org/TR/sparql11-query/#evaluation
+        return TriBool::Err;
+    }
+
+    if (handle_ == other.handle_) {
+        return TriBool::True;
+    }
+
+    if (handle_.type() != other.handle_.type()) {
+        return TriBool::Err;
     }
 
+    using storage::identifier::RDFNodeType;
+    switch (handle_.type()) {
+        case RDFNodeType::Literal: {
+            return Literal{handle_}.eq(Literal{other.handle_});
+        }
+        case RDFNodeType::IRI: {
+            return handle_.iri_backend() == other.handle_.iri_backend();
+        }
+        case RDFNodeType::BNode: {
+            return handle_.bnode_backend() == other.handle_.bnode_backend();
+        }
+        case RDFNodeType::Variable: {
+            return handle_.variable_backend() == other.handle_.variable_backend();
+        }
+        default: {
+            assert(false); // unreachable
+            return TriBool::Err;
+        }
+    }
+}
+
+std::partial_ordering Node::compare_impl(Node const &other) const noexcept {
+    if (null() || other.null()) {
+        // "Apart from BOUND, COALESCE, NOT EXISTS and EXISTS, all functions and operators operate on RDF Terms and will produce a type error if any arguments are unbound."
+        // - https://www.w3.org/TR/sparql11-query/#evaluation
+        return std::partial_ordering::unordered;
+    }
+
+    if (handle_ == other.handle_) {
+        return std::partial_ordering::equivalent;
+    }
+
+    if (handle_.type() != other.handle_.type() || handle_.type() != storage::identifier::RDFNodeType::Literal) {
+        // mismatched node types are not comparable
+        // and nodes other than literals are not comparable with <,<=,>,>=
+        return std::partial_ordering::unordered;
+    }
+
+    return Literal{handle_}.compare(Literal{other.handle_});
+}
+
+std::weak_ordering Node::order(Node const &other) const noexcept {
+    // first, check for null nodes
+    // "Apart from BOUND, COALESCE, NOT EXISTS and EXISTS, all functions and operators operate on RDF Terms and will produce a type error if any arguments are unbound."
+    // - https://www.w3.org/TR/sparql11-query/#evaluation
     if (this->null() && other.null()) {
         return this->handle_.type() <=> other.handle_.type();
     }
-
-    // unbound
     if (this->null()) {
-        return std::strong_ordering::less;
-    } else if (other.null()) {
-        return std::strong_ordering::greater;
+        return std::weak_ordering::less;
+    }
+    if (other.null()) {
+        return std::weak_ordering::greater;
+    }
+
+    if (this->handle_ == other.handle_) {
+        return std::weak_ordering::equivalent;
     }
```

Comment:
I would reorder it a bit

```suggestion
    if (this->handle_ == other.handle_) {
        return std::weak_ordering::equivalent;
    }
    if (this->handle_.null() || other.handle_.null()) {
        return this->handle_.null()
            ? std::weak_ordering::less
            : std::weak_ordering::greater;
    }
```

Reply:
Thats very subly wrong, it does not order null nodes of different types correctly

Reply:
```c++
std::weak_ordering Node::order(Node const &other) const noexcept {
    // first, check for null nodes
    // "Apart from BOUND, COALESCE, NOT EXISTS and EXISTS, all functions and operators operate on RDF Terms and will produce a type error if any arguments are unbound."
    // - https://www.w3.org/TR/sparql11-query/#evaluation
    if (this->null() && other.null()) {
        return this->handle_.type() <=> other.handle_.type();
    }

    if (this->handle_ == other.handle_) {
        return std::weak_ordering::equivalent;
    }
    if (this->handle_.null() || other.handle_.null()) {
        return this->handle_.null()
            ? std::weak_ordering::less
            : std::weak_ordering::greater;
    }
```